### ISSUE TITLE
Project builds fails with Java 9 or more (URLClassLoader issue)

### ISSUE THREAD

**Issue by Glodenox** [xspec/xspec-maven-plugin-1#46]
As mentioned in the [Java 9 release notes](https://bugs.java.com/bugdatabase/view_bug.do?bug_id=8183237), it is no longer possible to rely on the application class loaders to be an instance of `URLClassLoader`:

> The application class loader is no longer an instance of `java.net.URLClassLoader` (an implementation detail that was never specified in previous releases). Code that assumes that `ClassLoader::getSytemClassLoader` returns a `URLClassLoader` object will need to be updated. Note that Java SE and the JDK do not provide an API for applications or libraries to dynamically augment the class path at run-time.

classpath-utils [seems to rely on this](https://github.com/cmarchand/classpath-utils/blob/master/src/main/java/top/marchand/java/classpath/utils/ClasspathUtils.java#L118), and therefore any builds of this plugin on Java 9 or higher currently seem to fail. I wasn't certain whether I needed to add an issue there or here as I think a change would be needed in both projects to fix this.

This is the stack trace I am seeing:
```
io.xspec.maven.xspecMavenPlugin.utils.XSpecPluginException: while creating catalogBuilder
        at io.xspec.maven.xspecMavenPlugin.utils.CatalogWriter.<init>(CatalogWriter.java:64)
        at io.xspec.maven.xspecMavenPlugin.utils.CatalogWriter.<init>(CatalogWriter.java:68)
        at io.xspec.maven.xspecMavenPlugin.utils.XmlStuff.buildUriResolver(XmlStuff.java:249)
        at io.xspec.maven.xspecMavenPlugin.utils.XmlStuff.<init>(XmlStuff.java:164)
        at io.xspec.maven.xspecMavenPlugin.XSpecRunner.init(XSpecRunner.java:172)
        at io.xspec.maven.xspecMavenPlugin.XSpecRunnerTest.getNewRunner(XSpecRunnerTest.java:143)
        at io.xspec.maven.xspecMavenPlugin.XSpecRunnerTest.getNewRunner(XSpecRunnerTest.java:148)
        at io.xspec.maven.xspecMavenPlugin.XSpecRunnerTest.processXsltXspecTest(XSpecRunnerTest.java:86)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
[omitted further JUnit and Surefire stack entries]
Caused by: top.marchand.java.classpath.utils.ClasspathException: classloader is not a URL classloader : jdk.internal.loader.ClassLoaders$AppClassLoader
        at top.marchand.java.classpath.utils.ClasspathUtils.getClasspathElements(ClasspathUtils.java:129)
        at top.marchand.java.classpath.utils.ClasspathUtils.<init>(ClasspathUtils.java:61)
        at io.xspec.maven.xspecMavenPlugin.utils.CatalogWriter.<init>(CatalogWriter.java:62)
```

**Comment by cmarchand** [comment]
This concerns only Saxon extension functions load, probably used only by one client. We are waiting for a service mecanism from Saxon for this, but it is required by this client.
I'm gonna work on this. Thanks for pointing this out.

**Comment by Glodenox** [comment]
Thanks for looking into it! All in all it's not a major issue from my personal point of view as I'm still able to use the plugin in projects that use Java 9+. But for the future it would obviously be best that the plugin can also be built on newer versions of Java.

**Comment by cmarchand** [comment]
This is not a simple task, as we inspect classpath to generate the catalog. And all this stuff has changed with java 9.
An possibility from @jmdoudoux could be :
```
package fr.jmdoudoux.dej.java;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.lang.module.ModuleReader;
import java.lang.module.ResolvedModule;

public class ModuleScan {

  public static void main(String[] args) {
    ModuleLayer.boot()
               .configuration()
               .modules()
               .stream()
               .map(ResolvedModule::reference)
               .forEach(moduleref -> {
                 System.out.println("**" + moduleref.descriptor()
                                                    .name());
                 try (ModuleReader reader = moduleref.open()) {
                   reader.list()
                         .filter(s -> s.contains(".properties"))
                         .forEach(System.out::println);
                 } catch (IOException ioe) {
                   throw new UncheckedIOException(ioe);
                 }
               });
   }
}
```

**Comment by cmarchand** [comment]
We also need to have a module-compatible XSpec jar. Link to https://github.com/xspec/xspec/issues/938.

**Comment by cmarchand** [comment]
Problem solved by changing the way XML resources are loaded.

They are now loaded from classpath, using `cp:/` protocol, and it is not necessary to inspect classpath to create catalog. As a consequence, technical catalog used for maven plugin implementation is now empty.

This will be available in next release (probably a 2.04.xx)...
